migratekvdb/migration.go

```diff
@@ -0,0 +1,1140 @@
+
+		// Mark migration as complete.
+		err = m.markMigrationComplete()
+		if err != nil {
+			return err
+		}
+
+		// Update final state. This also makes sure the last
+		// chunk has a unique key.
+		m.migration.setFinalState()
+
+		// Write the migration state.
+		err = m.migration.write(m.cfg.MetaDB)
+		if err != nil {
+			return fmt.Errorf("failed to write migration "+
+				"state: %w", err)
+		}
+
+		if m.migration.resuming {
+			return fmt.Errorf("no keys were migrated, " +
+				"the migration state path was not " +
+				"found")
+		}
+
+		elapsed := time.Since(m.migration.persistedState.StartTime)
+
+		m.cfg.Logger.Infof("Migration completed in %v (since "+
+			"inital start)", elapsed)
+
+		return nil
+	}
+
+}
+
+// processChunk processes a single chunk of the migration by walking through
+// the nested bucket structure and migrating the key-value pairs up to the
+// specified chunk size.
+func (m *Migrator) processChunk(ctx context.Context,
+	sourceTx kvdb.RTx, targetTx kvdb.RwTx) error {
+
+	// We start the iteration by looping through the root buckets.
+	err := sourceTx.ForEachBucket(func(name []byte) error {
+		sourceRootBucket := sourceTx.ReadBucket(name)
+		if sourceRootBucket == nil {
+			return fmt.Errorf("source root bucket not found for "+
+				"key %x", name)
+		}
+
+		// We create the root bucket if it does not exist.
+		targetRootBucket, err := targetTx.CreateTopLevelBucket(
+			name,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to create target root "+
+				"bucket: %w", err)
+		}
+
+		if !m.migration.resuming {
+			m.cfg.Logger.Infof("Migrating root bucket: %v",
+				loggableKeyName(name))
+
+			// Increase the number of processed buckets.
+			m.migration.persistedState.ProcessedBuckets++
+		}
+
+		// Start with the root bucket name as the initial path.
+		initialPath := NewBucketPath([][]byte{name})
+		err = m.migrateBucket(
+			ctx, sourceRootBucket, targetRootBucket, initialPath,
+		)
+
+		return err
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// migrateBucket migrates a bucket from the source database to the target
+// database.
+func (m *Migrator) migrateBucket(ctx context.Context,
+	sourceB kvdb.RBucket, targetB kvdb.RwBucket,
+	path BucketPath) error {
+
+	m.logMigrationProgress(path)
+
+	// Helper variables to shorten the names.
+	currentPath := m.migration.persistedState.CurrentBucketPath
+	lastUnprocessedKey := m.migration.persistedState.LastUnprocessedKey
+
+	var alreadyMigrated bool
+	var fastForwardBucket []byte
+
+	// We skip the bucket if we already migrated it or fast-forward to the
+	// next bucket if we can.
+	switch m.migration.resuming {
+	case true:
+		alreadyMigrated, fastForwardBucket = checkBucketMigrated(
+			currentPath, path,
+		)
+
+		if alreadyMigrated {
+			m.cfg.Logger.Debugf("Skipping already migrated "+
+				"bucket: %v", path)
+
+			return nil
+		}
+
+	case false:
+		if err := targetB.SetSequence(sourceB.Sequence()); err != nil {
```

Comment:
I think we could spare some extra sequence keys if we only set the sequence if the source has a sequence (meaning non-zero). There're a lot of subbuckets, so a lot of extra sequences are added this way (as in sqlite/postgres it's an actual stored key).

Reply:
true, essentially we were setting them here to 0 rather than NULL, not sure if we are sparing here some sequence numbers but we keep them at `Null` if they are not set.

Reply:
Yeah we do as if we call `SetSequence`, that'll add extra keys to the keyspace which will be stored (inflating indexes etc etc).